### PR TITLE
Stop water from spreading over water.

### LINKED ISSUE #3529: Water flowing into water should not fan out
related to #2767.

Waterfalls that are falling down into another pool of water should not cause fanning out of the water. This creates lots of new flowing water nodes that are completely against what water as a liquid behaves like in reality. Maybe some splashing can be seen, but water mostly absorbs all the liquid without disturbing the area as it does right now.

![before adding waterfall]
(http://i.imgur.com/CStVUGj.png)

![after adding waterfall]
(http://i.imgur.com/NZjANlk.png)

expected bahavior: water falling down onto a liquid node just stops right there. nothing more.

### PR DESCRIPTION
**Pull request by red-001**
fixes #3529 

before:

screenshot_20160206_111247
After:
screenshot_20160206_110658

### CODE REVIEW

**Comment by ObaniGemini** [comment]
:+1:

File: src/map.cpp [hunk 1]
```diff
@@ -1744,6 +1744,9 @@ void Map::transformLiquids(std::map<v3s16, MapBlock*> & modified_blocks)
 					break;
 				case LIQUID_SOURCE:
 					// if this node is not (yet) of a liquid type, choose the first liquid type we encounter
+					if (nb.t == NEIGHBOR_LOWER){
```
Comment (nerzhul) [inline src/map.cpp]: missing space

**Comment by C1ffisme** [comment]
:+1: 

Thank you! I don't know how many waterfalls, waterslides, etc. have been ruined by this bug.

**Comment by paramat** [comment]
Will review, as i've been working on this code.

**Comment by paramat** [comment]
This is exactly what i would have tried, using the 'flowing down' flag.
See 'case LIQUID_FLOWING:' below and line 1768, perhaps this code should be in the 'else {' block?
I'll try to think it through later.
Also, when the code is one line there's no need for curly brackets:
```
if (nb.t == NEIGHBOR_LOWER)
flowing_down = true;
```

**Comment by paramat** [comment]
Yes the lines you added should be added as in the 'case LIQUID_FLOWING:' block below.
So add ..
```
						if (nb.t == NEIGHBOR_LOWER)
							flowing_down = true;
```
.. in the '} else {' block after ..
```
					} else {
						// Do not count bottom source, it will screw things up
						if(dirs[i].Y != -1)
							sources[num_sources++] = nb;
```

**Comment by paramat** [comment]
Hang on, there may be a reason for water to spread over water sources, this is what creates the water slopes on rivers, if it didn't then all rivers would become cubes-only with no sloping areas of flowing water, perhaps you could test that?

**Comment by red-001** [comment]
@paramat  With this PR water does spread out on to water sources but only if it flows from the side of the water source and only by one node. I should have phrased the title differently. Here is a screenshot to show what I mean
screenshot_20160208_084209

**Comment by kilbith** [comment]
Tested this PR in Valleys Mapgen. Valid point raised by @paramat : water flowing inside rivers is now either : spread over 1 node only with a 45° sloping, or no spread at all.

So obviously :-1:

screenshot_20160208_152543

screenshot_20160208_155116

**Comment by nerzhul** [comment]
:-1: valid point. Valleys mapgen is now part of core and use this feature

**Comment by red-001** [comment]
@kilbith  I couldn't find any cases of water not spreading at all.

**Comment by ObaniGemini** [comment]
Maybe make it a minetest.conf server parameter too ?

**Comment by paramat** [comment]
> Here is a screenshot to show what I mean

Note that liquid will spread on liquid if the 2 are different types of water.
The important point is river slopes, so closing. Thanks for the test.

**Comment by red-001** [comment]
@paramat the code doesn't care about liquid type other then flowing vs source ......

**Comment by red-001** [comment]
Personally I don't find slopes that are only 45° to be a problem but I will add a check that only set's flowing_down to true if the node above it is water.

**Comment by red-001** [comment]
@paramat  @nerzhul fix for the problem kilbith found: https://github.com/red-001/minetest/commit/651ba083162e6c5465e7bc02e36d6c96758ee0ce Could this PR be reopened?

**Comment by paramat** [comment]
> the code doesn't care about liquid type other then flowing vs source

It does if it's coded correctly as i detailed above.

I'm still not keen on this, it's getting complex for something that is a minor issue, i don't mind it spreading out. When a river or stream of water jumps down 2 nodes it would be preferable to still have it spread.
River slopes longer than 1 node are important and often used in lua mapgens.

**Comment by red-001** [comment]
So water that act's as if `water_source` is a solid node is a feature? I know minetest is made out of voxels and blocks can just float in the air but water seems to follow the laws of physics for the most part. 
>I'm still not keen on this, it's getting complex for something that is a minor issue

I would say water spreading only by one node with a 45 degree slope is a lot less of an issue. 

> When a river or stream of water jumps down 2 nodes it would be preferable to still have a slope.

Is unrealistic water really a feature?

>River slopes longer than 1 node are important and often used in lua mapgens.

 See  red-001@651ba08.

**Comment by paramat** [comment]
Your PR is not correct as you haven't moved the code as i explained above.

Minetest water is unrealistic in many ways because the code needs to be simple due to how many nodes are processed each cycle (up to 100000). So any additional code has to be strongly justified, i don't feel this spreading is enough of an issue to justify, even though i agree it looks a little weird.
Essentially we shouldn't be trying to fix minor water realism issues because there are many and the code would become too bloated.

**Comment by paramat** [comment]
Closing for reasons above.

**Comment by red-001** [comment]
@kilbith water freezing seems to happen when a map block gets unloaded or the server shuts down  before water has time to update.

**Pull request closed without merging**